pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -483,34 +483,39 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
 				}
 
 			case RouteType:
+				if discReq.ErrorDetail != nil {
+					adsLog.Warnf("ADS:RDS: ACK ERROR %v %s (%s) %v", peerAddr, con.ConID, con.modelNode.ID, discReq.String())
+					rdsReject.With(prometheus.Labels{"node": discReq.Node.Id, "err": discReq.ErrorDetail.Message}).Add(1)
+					totalXDSRejects.Add(1)
+					continue
+				}
 				routes := discReq.GetResourceNames()
+				if routes == nil && discReq.ResponseNonce != "" {
+					// There is no requirement that ACK includes routes. The test doesn't.
+					con.mu.Lock()
+					con.RouteNonceAcked = discReq.ResponseNonce
+					con.mu.Unlock()
+					continue
+				}
+				// routes and con.Routes are all empty, this is not an ack and will do nothing.
+				if len(routes) == 0 && len(con.Routes) == 0 {
+					continue
+				}
 
-				if len(routes) == len(con.Routes) || len(routes) == 0 {
-					// routes and con.Routes are all empty, this is not an ack and will do nothing.
-					if len(routes) == 0 && len(con.Routes) == 0 {
-						continue
-					}
-
+				if len(routes) == len(con.Routes) {
 					sort.Strings(routes)
 					sort.Strings(con.Routes)
 
-					if reflect.DeepEqual(con.Routes, routes) || len(routes) == 0 {
```

Comment:
Why did you remove the prometheus tracking of nacks ? That was pretty important.